### ISSUE TITLE
security group rule will change back to 0.0.0.0 by k8s after I changed it. (AWS ELB)

### ISSUE THREAD

**Issue by johnzheng1975** [kubernetes/kubernetes#64259]
security group rule will change back to 0.0.0.0 by k8s after I changed it. (AWS ELB)

*k8s 1.9
*kubespray 2.4
*AWS

Steps:
1 I created a service with Loadbalancer type
2 I find the security group of elb is open to worldwide 0.0.0.0
3 I changed the security group rule to my company ip xx.xx.xx.xx/8
4 After some time, I checked the security group rule, it is 0.0.0.0 again.
5 From cloud trail, it showed aws role (for aws cluster) changed it automatically.

Expect result:
I want to change and keep security group rule to xx.xx.xx.xx/8

Real result:
It always change back to 0.0.0.0.

Note:
For my other k8s cluster, it do not change it for some time.

**Comment by johnzheng1975** [comment]
@kubernetes/sig-bugs

**Comment by Bizzelicious** [comment]
Hi, you can use `loadBalancerSourceRanges` in the service manifest to whitelist IPs in your SG. With this option the IPs will be whitelisted when the service creates your load balancer. No need to whitelist afterwards.
```
kind: Service
apiVersion: v1
metadata:
  name: foo
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-cross-zone-load-balancing-enabled: "True"
spec:
  loadBalancerSourceRanges: 
    - x.x.x.x.x/32
    - y.y.y.y.y/32
  type: LoadBalancer
  selector:
    app: foo
  ports:
  - name: foo
    protocol: TCP
    port: 9000
    targetPort: 9000
```

**Comment by johnzheng1975** [comment]
I will not use this way until there is no other choice, because:
1   I need define the duplicate ip for each micro-service
2   The whitelist may change frequently, I need maintain and replace service frequently.
3   For each rule, I also like to add comment(whose ip it is), this do not support.

So, my question is
1  does "k8s change ip back to default ip (i.e. 0.0.0.0)" by design?
2  When it will change, is there any cron job in k8s for this?
3  which k8s code for this function, I did not find it.
https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/aws/aws.go

Thanks a lot.

**Comment by johnzheng1975** [comment]
Here is an answer from hubt, seems good.
yes it does. in general, if kubernetes see anything that is not it's desired config, it will attempt to change it so that it is the desired config.
that's the general controller paradigm that is use throughout kubernetes. so for something like this, you need to define the security groups to add to the elb in the service.
use something like this: https://kubernetes.io/docs/tasks/access-application-cluster/configure-cloud-provider-firewall/#restrict-access-for-loadbalancer-service
kubernetes.io
Configure Your Cloud Provider's Firewalls - Kubernetes
Configure Your Cloud Provider's Firewalls

it's not in the aws cloud provider, because it applies to all clouds

**Comment by Raffo** [comment]
A possible solution to that would be to be able to use a Security Group instead. There is a PR open here https://github.com/kubernetes/kubernetes/pull/62774 which would allow to do that. The pre-existing annotation on the ELB allows only to setup an *extra* Security Group via an annotation but does not remove the `0.0.0.0/0` from the LoadBalancer.

**Comment by johnzheng1975** [comment]
Since it is by design of k8s, I closed this issue.
However, it will be great let user add an annotation to decide whether sg need changed back or not.

I used the method suggested by Bizzelicious & hubt finally.
I still do not know which code will check the sg and change back continually.  Is it use "k8s job", or "cron job"? It will be appreciate if someone can show code to me, thanks in advance.
